radixdlt/src/integration/java/com/radixdlt/consensus/simulation/tests/epochs/RandomValidatorsTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * (C) Copyright 2020 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.consensus.simulation.tests.epochs;
+
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+
+import com.radixdlt.consensus.View;
+import com.radixdlt.consensus.simulation.SimulationTest;
+import com.radixdlt.consensus.simulation.SimulationTest.Builder;
+import com.radixdlt.consensus.simulation.TestInvariant.TestInvariantError;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.assertj.core.api.AssertionsForClassTypes;
+import org.assertj.core.api.Condition;
+import org.junit.Test;
+
+public class RandomValidatorsTest {
+	private static final int numNodes = 50;
+
+	private final Builder bftTestBuilder = SimulationTest.builder()
+		.pacemakerTimeout(5000)
+		.numNodes(numNodes)
+		.epochHighView(View.of(100))
+		.checkEpochHighView("epochHighView", View.of(100))
+		.checkSafety("safety")
+		.checkLiveness("liveness", 5000, TimeUnit.MILLISECONDS)
+		.checkNoTimeouts("noTimeouts")
+		.checkAllProposalsHaveDirectParents("directParents");
+
+	private static Function<Long, IntStream> randomEpochToNodesMapper(Function<Long, Random> randomSupplier) {
+		return epoch -> {
+			List<Integer> indices = IntStream.range(0, numNodes).boxed().collect(Collectors.toList());
+			Random random = randomSupplier.apply(epoch);
+			for (long i = 0; i < epoch; i++) {
+				random.nextInt(numNodes);
+			}
+			return IntStream.range(0, random.nextInt(numNodes) + 1)
+				.map(i -> indices.remove(random.nextInt(indices.size())));
+		};
+	}
+
+	private static Function<Long, IntStream> goodRandomEpochToNodesMapper() {
+		return randomEpochToNodesMapper(Random::new);
+	}
+
+	private static Function<Long, IntStream> badRandomEpochToNodesMapper() {
+		// random number generator which gives a different number per node
```

Comment:
Hmm, until this comment appeared, it was not at all obvious to me that this mapper is executed for each node, rather than the validator set as a whole.  I think it would be worth adding a comment to `epochToNodesMapper()` explaining this.